interpreter/src/core_functions.rs

```diff
@@ -0,0 +1,1153 @@
+// Copyright (C) 2019-2024 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
+use super::*;
+
+use leo_ast::{CoreFunction, Expression};
+use leo_errors::Result;
+use leo_span::{Span, Symbol};
+
+use snarkvm::prelude::{CastLossy as _, Network as _, TestnetV0, ToBits};
+
+use rand::Rng as _;
+use rand_chacha::ChaCha20Rng;
+
+pub trait CoreFunctionHelper {
+    fn pop_value_impl(&mut self) -> Option<Value>;
+
+    fn pop_value(&mut self) -> Result<Value> {
+        match self.pop_value_impl() {
+            Some(v) => Ok(v),
+            None => {
+                Err(InterpreterHalt::new("value expected - this may be a bug in the Leo interpreter".to_string())
+                    .into())
+            }
+        }
+    }
+
+    fn set_block_height(&mut self, _height: u32) {}
+
+    fn lookup_mapping(&self, _program: Option<Symbol>, _name: Symbol) -> Option<&HashMap<Value, Value>> {
+        None
+    }
+
+    fn lookup_mapping_mut(&mut self, _program: Option<Symbol>, _name: Symbol) -> Option<&mut HashMap<Value, Value>> {
+        None
+    }
+
+    fn rng(&mut self) -> Option<&mut ChaCha20Rng> {
+        None
+    }
+}
+
+impl CoreFunctionHelper for Vec<Value> {
+    fn pop_value_impl(&mut self) -> Option<Value> {
+        self.pop()
+    }
+}
+
+pub fn evaluate_core_function(
+    helper: &mut dyn CoreFunctionHelper,
+    core_function: CoreFunction,
+    arguments: &[Expression],
+    span: Span,
+) -> Result<Option<Value>> {
+    macro_rules! apply {
```

Comment:
In hopes of minimizing divergence from snarkVM, would an alternate design leveraging the `evaluate` methods of the instructions work? Roughly speaking, you'd create a type that implements the necessary traits like `StackMatches`, `RegistersLoad`, and then just invoke the implementation directly.

https://github.com/ProvableHQ/snarkVM/blob/4a22b0ada6e293bfcaa2711e4d61dc309a3649d3/synthesizer/program/src/logic/instruction/operation/hash.rs#L202

Reply:
Agreed , could we throw it in the issues for tracking purposes?